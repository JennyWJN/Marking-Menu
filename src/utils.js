/**
 * @param {number} a the dividend
 * @param {number} n the divisor
 * @return {number} The modulo of `a` over `n` (% is not exactly modulo but remainder).
 */
export const mod = (a, n) => (a % n + n) % n;

/**
 * @param {number} alpha a first angle (in degrees)
 * @param {number} beta a second angle (in degrees)
 * @return {number} The (signed) delta between the two angles (in degrees).
 */
export const deltaAngle = (alpha, beta) => mod(beta - alpha + 180, 360) - 180;

/**
 * @param {String} [prefix] - A prefix to append before log messages.
 * @return {Observable} An observer logging next, errors and complete "events".
 *                      Handy to debug observable.
 */
export const logObservable = prefix => {
  const fixedPrefix = prefix ? `${prefix} ` : '';
  return {
    next(e) {
      // eslint-disable-next-line no-console
      console.log(`${fixedPrefix}next`, e);
    },
    error(e) {
      // eslint-disable-next-line no-console
      console.error(`${fixedPrefix}error`, e);
    },
    complete() {
      // eslint-disable-next-line no-console
      console.log(`${fixedPrefix}complete`);
    }
  };
};

/**
 * Calculate the euclidean distance between two
 * points.
 *
 * @param {List<number>} point1 - The first point
 * @param {List<number>} point2 - The second point
 * @return {number} The distance between the two points.
 */
export const dist = (point1, point2) => {
  const sum = point1.reduce((acc, x1i, i) => {
    const x2i = point2[i];
    return acc + (x2i - x1i) ** 2;
  }, 0);
  return Math.sqrt(sum);
};
