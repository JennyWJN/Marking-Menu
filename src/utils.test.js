import test from 'ava';
import { deltaAngle, mod, dist, angle, findMaxEntry } from './utils';

test('`mod` returns a positive modulo', t => {
  t.is(mod(3, 10), 3);
  t.is(mod(-3, 10), 7);
  t.is(mod(13, 10), 3);
  t.is(mod(-13, 10), 7);
});

test('`deltaAngle` properly returns the delta between two angles', t => {
  t.is(deltaAngle(40, 50), 10);

  t.is(deltaAngle(40, 50 + 360), 10);
  t.is(deltaAngle(40, 50 - 360), 10);
  t.is(deltaAngle(40 + 360, 50), 10);
  t.is(deltaAngle(40 - 360, 50), 10);

  t.is(deltaAngle(50, 40), -10);

  t.is(deltaAngle(50, 40 + 360), -10);
  t.is(deltaAngle(50, 40 - 360), -10);
  t.is(deltaAngle(50 + 360, 40), -10);
  t.is(deltaAngle(50 - 360, 40), -10);
});

test('`dist` returns the euclidean distance between two vectors', t => {
  t.is(dist([0, 0], [0, 0]), 0);
  t.is(dist([42, 11], [42, 11]), 0);

  t.is(dist([0, 0], [0, 1]), 1);
  t.is(dist([1, 0], [0, 0]), 1);

  t.is(dist([5, 10], [20, 4]), Math.sqrt((5 - 20) ** 2 + (10 - 4) ** 2));
  t.is(dist([20, 4], [5, 10]), Math.sqrt((5 - 20) ** 2 + (10 - 4) ** 2));

  t.is(dist([42, 11, 5, 8, 0], [42, 11, 5, 8, 0]), 0);
  t.is(dist([1, 1, 0, 0, 1], [0, 0, 1, 1, 0]), Math.sqrt(5));
  t.is(
    dist([1, 2, 3, 7, 8], [4, 5, 6, 9, 10]),
    Math.sqrt(
      (1 - 4) ** 2 + (2 - 5) ** 2 + (3 - 6) ** 2 + (7 - 9) ** 2 + (8 - 10) ** 2
    )
  );
});

test('`angle` calculate an angle from three points', t => {
  t.is(angle([5, 1], [1, 1], [1, 20]), 90, 1e-9);
  t.is(angle([5, 1], [1, 1], [1, -20]), 90, 1e-9);
  t.is(angle([5, 1], [1, 1], [110, 1]), 0, 1e-9);
  t.is(angle([5, 5], [1, 5], [-10, 5]), 180, 1e-9);
  t.is(angle([5, 5], [1, 1], [5, 1]), 45, 1e-9);
  t.is(angle([5, 5], [1, 1], [-10, -10]), 180, 1e-9);
});

test('`findMaxEntry` return an entry with the item that raised the highest number', t => {
  t.deepEqual(
    findMaxEntry(
      [{ l: 2 }, { l: 0 }, { l: 10 }, { l: 5 }],
      (a, b) => b.l - a.l
    ),
    [2, { l: 10 }]
  );
  t.deepEqual(
    findMaxEntry(
      [{ l: 2 }, { l: 0 }, { l: 10 }, { l: 5 }],
      (a, b) => a.l - b.l
    ),
    [1, { l: 0 }]
  );
});
