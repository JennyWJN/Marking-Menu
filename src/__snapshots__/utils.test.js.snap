// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strToHTML  parses a string as a DOM Element 1`] = `
Array [
  <ul
    style="position: absolute;"
  >
    
          
    <li>
      That thing
    </li>
    
          
    <li>
      That other thing
    </li>
    
        
  </ul>,
  <svg
    height="100"
    viewBox="0 0 3 2"
    width="150"
    xmlns="http://www.w3.org/2000/svg"
  >
    
          
    <rect
      fill="#008d46"
      height="2"
      width="1"
      x="0"
    />
    
          
    <rect
      fill="#ffffff"
      height="2"
      width="1"
      x="1"
    />
    
          
    <rect
      fill="#d2232c"
      height="2"
      width="1"
      x="2"
    />
    
        
  </svg>,
  <div
    class="that-thing"
  >
    I exist too
  </div>,
]
`;
